proposals/0052-fonts.md

```diff
@@ -0,0 +1,369 @@
+
+## Astro config
+
+### Overview
+
+The goal is to have a config that starts really simple for basic usecases, but can also be complex for advanced usecases. Here's an example of basic config:
+
+```js
+import { defineConfig } from "astro/config";
+
+export default defineConfig({
+  fonts: {
+    families: ["Roboto", "Lato"],
+  },
+});
+```
+
+That would get fonts from [Google Fonts](https://fonts.google.com/) with sensible defaults.
+
+Here's a more complex example:
+
+```js
+import { defineConfig, fontProviders } from "astro/config";
+import { myCustomFontProvider } from "./provider";
+
+export default defineConfig({
+  fonts: {
+    providers: [
+      fontProviders.adobe({ apiKey: process.env.ADOBE_FONTS_API_KEY }),
+      myCustomFontProvider(),
+    ],
+    defaults: {
+      provider: "adobe",
+      weights: [200, 700],
+      styles: ["italic"],
+      subsets: [
+        "cyrillic-ext",
+        "cyrillic",
+        "greek-ext",
+        "greek",
+        "vietnamese",
+        "latin-ext",
+        "latin",
+      ],
+    },
+    families: [
+      "Roboto",
+      {
+        name: "Lato",
+        provider: "google",
+        weights: [100, 200, 300],
+      },
+      {
+        name: "Custom",
+        provider: "local",
+        src: ["./assets/fonts/Custom.woff2"],
+      },
+    ],
+  },
+});
+```
+
+### Providers
+
+#### Definition
+
+A provider allows to retrieve font faces data from a font family name from a given CDN or abstraction. It's a [unifont](https://github.com/unjs/unifont) provider.
+
+#### Built-in providers
+
+##### Google
+
+This is the default, and it's not configurable. Given the amount of fonts it supports by, it sounds like a logic choice. Note that the default can be customized for more advanced usecases.
+
+```js
+export default defineConfig({
+  fonts: {
+    families: ["Roboto"],
+  },
+});
+```
+
+```js
+export default defineConfig({
+  fonts: {
+    defaults: {
+      provider: "local",
+    },
+    families: [
+      {
+        name: "Roboto",
+        provider: "google",
+      },
+    ],
+  },
+});
+```
+
+##### Local
+
+This provider, unlike all the others, requires paths to fonts relatively to the root.
+
+```js
+import { defineConfig, fontProviders } from "astro/config";
+
+export default defineConfig({
+  fonts: {
+    families: [
+      {
+        name: "Custom",
+        provider: "local",
+        src: ["./assets/fonts/Custom.woff2"],
+      },
+    ],
+  },
+});
+```
+
+#### Opt-in providers
+
+Other unifont providers are exported from `astro/config`.
+
+```js
+import { defineConfig, fontProviders } from "astro/config";
+
+export default defineConfig({
+  fonts: {
+    providers: [
+      fontProviders.adobe({ apiKey: process.env.ADOBE_FONTS_API_KEY }),
+    ],
+    // ...
+  },
+});
+```
+
+#### Why this API?
+
+1. **Coherent API**: a few things in Astro are using this pattern, namely integrations and vite plugins. It's simple to author as a library author, easy to use as a user
```

Comment:
I disagree with this statement, this API is used **only** for integrations and vite plugins. 

For example: middleware, actions, and content collections (we could consider them an exception though) don't follow this pattern. They use a **known file** to collect users' APIs. Same thing for `@astrojs/db` (but that's integration).

I think that our configuration should inject as less runtime as possible, use good defaults, and maybe the custom providers could follow the same pattern we use with actions and middleware.

Reply:
Image services, which are probably the most similar thing do work this way

Reply:
Probably I missed something, please correct me if I misunderstood what you mean. Based on this docs page (https://docs.astro.build/en/reference/image-service-reference/#user-configuration), the image service uses an entry point API, which is different from the one proposed.

Reply:
Ah, apologies, I misunderstood what you meant. Thought this was mostly about where they're configured.

Image services do use an entrypoint, but they're configured in the Astro config, when they have settings.

Reply:
The main advantage of having objects in the config directly for providers is typing support, which imo is super important

Reply:
Alright so I guess you'd prefer such an API, am I right?
```js
defineConfig({
	fonts: {
		providers: [fontProviders.adobe({ apiKey: "..." })]
		// equal to { name: "adobe", entrypoint: "astro/assets/fonts/adobe", config: { apiKey: "..." } }
	}
})
```
I guess that's doable